contracts/RestrictedNORI.sol

```diff
@@ -829,52 +826,52 @@ contract RestrictedNORI is
     uint256 removalId,
     uint256 amount
   ) internal {
-    // slither-disable-next-line calls-loop
+    // slither-disable-next-line calls-loop choose to get the project id from the removal contract
     uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
     Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
     if (!schedule.exists) {
       revert NonexistentSchedule({scheduleId: scheduleId});
     }
-    uint256 quantityToRevoke;
-    // amount of zero indicates revocation of all remaining tokens.
-    if (amount > 0) {
-      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
-        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
-      }
-      quantityToRevoke = amount;
-    } else {
-      quantityToRevoke = revocableQuantityForSchedule(scheduleId);
+    if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+      revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
     }
 
+    // amount of zero indicates revocation of all remaining tokens.
+    uint256 quantityToRevoke = amount > 0
+      ? amount
+      : revocableQuantityForSchedule(scheduleId);
+
     // burn correct proportion from each token holder
-    EnumerableSetUpgradeable.AddressSet storage tokenHolders = schedule
-      .tokenHolders;
+    address[] memory tokenHoldersLocal = schedule.tokenHolders.values();
     uint256[] memory quantitiesToBurnForHolders = new uint256[](
-      tokenHolders.length()
+      tokenHoldersLocal.length
     );
 
+    // Calculate the final holder's quantity to revoke by subtracting the sum of other quantities
+    // from the desired total to revoke, thus avoiding any precision rounding errors from affecting
+    // the total quantity revoked by up to several wei.
     uint256 cumulativeQuantityToBurn = 0;
-    for (uint256 i = 0; i < (tokenHolders.length() - 1); i++) {
+    for (uint256 i = 0; i < (tokenHoldersLocal.length - 1); i++) {
```

Comment:
`_quantityToRevokePerTokenHolder` is running `balanceOf` in each iteration and this could be optimized (maybe not in this pr?) to just accept a `tokenHolderBalances[]` argument which has a value assigned outside of this loop using a single batchBalanceOf query

Reply:
me neither. In the case of calling balanceOf another contract, it's definitely less overhead, but not sure in the case of super. Could you leave a todo to check?